content/courses/advanced-ada/parts/data_types/records.rst

```diff
@@ -2918,20 +2928,684 @@ In this example, :ada:`Arr_Sub` and :ada:`Rec_Sub` are unconstrained subtypes.
+           private;
+
+       procedure Turn_Off (D : in out Device);
+
+       procedure Turn_On (D : in out Device);
+
+       type Device_Info is
+       record
+          V : Float;
+       end record;
+
+       function Read_Info (D : Device)
+                           return Device_Info;
+
+    private
+
+       type Device (State : Device_State := Off) is
+       record
+          case State is
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+          end case;
+       end record;
+
+       Device_Off : constant Device :=
+                      (State => Off);
+
+       Device_On  : constant Device :=
+                      (State => On,
+                       others => <>);
+
+    end Devices;
+
+    package body Devices is
+
+       procedure Turn_Off (D : in out Device) is
+       begin
+          D := Device_Off;
+       end Turn_Off;
+
+       procedure Turn_On (D : in out Device) is
+       begin
+          D := Device_On;
+       end Turn_On;
+
+       function Read_Info (D : Device)
+                           return Device_Info is
+         (D.Info);
+
+    end Devices;
+
+Let's create a test application called :ada:`Show_Device` that makes use of
+this device by turning it on and off, and by retrieving information from it:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+    :class: ada-run-expect-failure
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D : Device;
+       I : Device_Info;
+    begin
+       Turn_On (D);
+       I := Read_Info (D);
+
+       Turn_Off (D);
+
+       --  The following call raises
+       --  an exception at runtime
+       --  because D is turned off.
+       I := Read_Info (D);
+    end Show_Device;
+
+In this example, by using the variant part, we're preventing that information
+retrieved by an inappropriate call to the :ada:`Read_Info` function is used
+elsewhere in the application. In fact, if the device is turned off, a call to
+:ada:`Read_Info` raises the :ada:`Constraint_Error` exception because the
+:ada:`Info` component isn't accessible. We see that effect in the
+:ada:`Show_Device` procedure: the call to :ada:`Read_Info` *fails* (by raising
+an exception) when the device has just been turned off.
+
+To avoid exceptions at runtime, we must check the device's state before
+calling :ada:`Read_Info`:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D : Device;
+       I : Device_Info;
+    begin
+       Turn_On (D);
+
+       if D.State = On then
+          I := Read_Info (D);
+       end if;
+
+       Turn_Off (D);
+
+       if D.State = On then
+          I := Read_Info (D);
+       end if;
+    end Show_Device;
+
+Now, no exception is being raised anymore, as we only retrieve information from
```

Comment:
Similarly "Now, no exception is being raised anymore," 
could be 
"Now, no exception is raised,"

Reply:
Thanks!